lib/location/method_gnss.c

```diff
@@ -0,0 +1,839 @@
+	if (err) {
+		LOG_ERR("A-GPS data processing failed, error: %d", err);
+		return;
+	}
+
+	LOG_DBG("A-GPS data processed");
+}
+#endif // defined(CONFIG_NRF_CLOUD_MQTT)
+#endif // defined(CONFIG_NRF_CLOUD_AGPS)
+
+#if defined(CONFIG_NRF_CLOUD_PGPS) && !defined(CONFIG_NRF_CLOUD_MQTT)
+static void method_gnss_pgps_request_work_fn(struct k_work *item)
+{
+	struct nrf_cloud_rest_context rest_ctx = {
+		.connect_socket = -1,
+		.keep_alive = false,
+		.timeout_ms = CONFIG_NRF_CLOUD_REST_RECV_TIMEOUT * MSEC_PER_SEC,
+		.auth = jwt_buf,
+		.rx_buf = rest_api_recv_buf,
+		.rx_buf_len = sizeof(rest_api_recv_buf),
+		.fragment_size = 0
+	};
+
+	int err = nrf_cloud_jwt_generate(0, jwt_buf, sizeof(jwt_buf));
+
+	if (err) {
+		LOG_ERR("Failed to generate JWT, error: %d", err);
+		return;
+	}
+
+	struct nrf_cloud_rest_pgps_request request = {
+		.pgps_req = &pgps_request
+	};
+
+	err = nrf_cloud_rest_pgps_data_get(&rest_ctx, &request);
+	if (err) {
+		nrf_cloud_pgps_request_reset();
+		LOG_ERR("nRF Cloud P-GPS request failed, error: %d", err);
+		return;
+	}
+
+	LOG_DBG("P-GPS data requested");
+
+	err = nrf_cloud_pgps_process(rest_ctx.response, rest_ctx.response_len);
+	if (err) {
+		nrf_cloud_pgps_request_reset();
+		LOG_ERR("P-GPS data processing failed, error: %d", err);
+		return;
+	}
+
+	LOG_DBG("P-GPS data processed");
+}
+#endif
+
+bool method_gnss_agps_required(struct nrf_modem_gnss_agps_data_frame *request)
+{
+	int type_count = 0;
+
+#if !defined(CONFIG_NRF_CLOUD_PGPS)
+	/* If P-GPS is enabled, use predicted ephemeris to save power instead of requesting them
+	 *  using A-GPS.
+	 */
+	if (request->sv_mask_ephe) {
+		type_count++;
+	}
+	if (request->sv_mask_alm) {
+		type_count++;
+	}
+#endif
+	if (request->data_flags & NRF_MODEM_GNSS_AGPS_GPS_UTC_REQUEST) {
+		type_count++;
+	}
+	if (request->data_flags & NRF_MODEM_GNSS_AGPS_KLOBUCHAR_REQUEST) {
+		type_count++;
+	}
+	if (request->data_flags & NRF_MODEM_GNSS_AGPS_NEQUICK_REQUEST) {
+		type_count++;
+	}
+	if (request->data_flags & NRF_MODEM_GNSS_AGPS_SYS_TIME_AND_SV_TOW_REQUEST) {
+		type_count++;
+	}
+	if (request->data_flags & NRF_MODEM_GNSS_AGPS_POSITION_REQUEST) {
+		type_count++;
+	}
+	if (request->data_flags &  NRF_MODEM_GNSS_AGPS_INTEGRITY_REQUEST) {
+		type_count++;
+	}
+
+	if (type_count == 0) {
+		LOG_DBG("No A-GPS data types requested");
+		return false;
+	} else {
+		return true;
+	}
+}
+
+#if defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_NRF_CLOUD_PGPS)
+static void method_gnss_request_assistance(void)
+{
+	int err = nrf_modem_gnss_read(&agps_request,
+				      sizeof(agps_request),
+				      NRF_MODEM_GNSS_DATA_AGPS_REQ);
+
+	if (err) {
+		LOG_WRN("Reading A-GPS req data from GNSS failed, error: %d", err);
+		return;
+	}
+
+	LOG_DBG("A-GPS request from modem (ephe: 0x%08x alm: 0x%08x flags: 0x%02x)",
+		agps_request.sv_mask_ephe,
+		agps_request.sv_mask_alm,
+		agps_request.data_flags);
+#if defined(CONFIG_LOCATION_METHOD_GNSS_AGPS_EXTERNAL)
```

Comment:
Should we do `if (IS_ENABLED(CONFIG_LOCATION_METHOD_GNSS_AGPS_EXTERNAL)`?

That's the style Robert has mentioned in earlier PR's for flagging inside functions.

Reply:
If possible, that's the preferred approach. I'm not sure if it'd be applicable in this particular case, `method_gnss_agps_ext_work` variable is complied out conditionally, so it'd be needed to modify other places as well. I leave it up to you if it's worth changing.